ml_tools/eolearn/ml_tools/TDigestTask.py

```diff
@@ -0,0 +1,178 @@
+"""
+The module provides an EOTask for the computation of a T-Digest representation of an EOPatch.
+
+Credits:
+Copyright (c) 2023 Michael Engel
+
+This source code is licensed under the MIT license found in the LICENSE
+file in the root directory of this source tree.
+"""
+from itertools import product
+from typing import Optional
+
+import numpy as np
+import tdigest as td
+
+from eolearn.core import EOPatch, EOTask, FeatureType
+from eolearn.core.types import FeaturesSpecification
+
+
+class TDigestTask(EOTask):
+    """
+    An EOTask to compute the T-Digest representation of a chosen feature of an EOPatch.
+    It integrates the [T-Digest algorithm by Ted Dunning](https://arxiv.org/abs/1902.04023) to efficiently \
+        compute quantiles of the underlying dataset into eo-learn.
+    The output features of the tasks may be merged to compute a representation of the complete dataset.
+    That enables quantile based normalisation or statistical analysis of datasets larger than RAM in EO.
+    """
+
+    def __init__(
+        self,
+        in_feature: FeaturesSpecification,
+        out_feature: FeaturesSpecification,
+        mode: Optional[str] = None,
+        pixelwise: Optional[bool] = None,
+    ):
+        """
+        :param in_feature: The input feature to compute the T-Digest representation for.
+        :param out_feature: The output feature where to save the T-Digest representation of the chosen feature.
+        :param mode: The mode to apply to the timestamps and bands.
+        - The 'standard' mode computes the T-Digest representation \
+            for each band accumulating timestamps.
+        - The 'timewise' mode computes the T-Digest representation \
+            for each band and timestamp of the chosen feature.
+        - The 'monthly' mode computes the T-Digest representation \
+            for each band accumulating the timestamps per month.
+        - The 'total' mode computes the total T-Digest representation \
+            of the whole feature accumulating all timestamps, bands and pixels \
+                - cannot be used with pixelwise=True.
+        :param pixelwise: Decider whether to compute the T-Digest representation accumulating pixels or per pixel. \
+            Cannot be used with mode='total'.
+        """
+
+        # check pixelwise parameter
+        if not pixelwise:
+            self.pixelwise = False
+        elif pixelwise and self.mode == "total":
+            raise ValueError("Total mode does not support pixelwise=True.")
+        else:
+            self.pixelwise = pixelwise
+
+        # set mode
+        if not mode:
+            self.mode = "standard"
+        else:
+            self.mode = mode
```

Comment:
This can likewise be solved by having `mode: str = "standard"` and `self.mode=mode`.

If you want to help out the users even more, you can import `Literal` from `typing_extensions` and have `mode: Literal["standard", "timewise", "monthly", "total"] = "standard"`. This way a type-checker can warn them about typos

Reply:
So you mean to support python 3.7? because from 3.8 ongoing, you could import it from typing itself, right?

Reply:
we'll probably be removing support for 3.7 in the next major release, until then we need the whole codebase to be 3.7 compliant. But you are indeed correct, in 3.8 and higher you can import it directly from `typing`.